spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java

```diff
@@ -320,7 +320,7 @@ private ClientSession initClientSession() throws IOException {
 		return clientSession;
 	}
 
-	private void initClient() throws IOException {
+	private synchronized void initClient() throws IOException {
```

Comment:
Synchronization has to be higher up - otherwise if `isSharedSession` is true, we could orphan a session; see code in `getSession`.

https://github.com/spring-projects/spring-integration/blob/f8fff8188622e4a9602cc46e430ab38776c58f31/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java#L286-L288

Reply:
If `isSharedSession` we do `this.sharedSessionLock.lock();`

Reply:
Oops - right; ignore.